Handle gluster mount options correctly

leelavg opened the pull request:
- Parse most of the gluster mount options
- If supplied options are incorrect log a message and proceed mounting without options

Fixes: #716

leelavg committed:
Handle gluster mount options correctly

Fixes: #716

Signed-off-by: Leela Venkaiah G <leelavg@thoughtexpo.com>

aravindavk commented:
Instead of revert, Change the code to use the options directly without `-o` So that most options will work fine. Split the Options by `,` then split by `=` as key and value. If value == "" then use as `--KEY` else use as `--KEY=VALUE`.

aravindavk commented on csi/volumeutils.py:
```diff
@@ -1135,6 +1135,19 @@ def mount_glusterfs_with_host(volname, mountpoint, hosts, options=None, is_clien
+    # if options:
+    #     opt_array = []
+    #     for opt in options.split(","):
+    #         if not opt or opt == "":
```
Use `opt.strip()` to compare as empty string. (Example: `-o "log-level= DEBUG,no-root-squash, "`)

aravindavk commented on csi/volumeutils.py:
```diff
@@ -1135,6 +1135,19 @@ def mount_glusterfs_with_host(volname, mountpoint, hosts, options=None, is_clien
+    #     opt_array = []
+    #     for opt in options.split(","):
+    #         if not opt or opt == "":
+    #             break
```
continue

aravindavk commented on csi/volumeutils.py:
```diff
@@ -1135,6 +1135,19 @@ def mount_glusterfs_with_host(volname, mountpoint, hosts, options=None, is_clien
+    #         if not opt or opt == "":
+    #             break
+    #         for k,v in opt.split("="):
+    #             if k == "log-level":
```
remove the custom check for now

aravindavk replied:
```
opt_array.append(f"--{k.strip()}")
if v.strip() != "":
    opt_array.append(v.strip())
```

aravindavk reviewed (commented):
Added suggestion in the code for quick fix

leelavg commented:
> Instead of revert, Change the code to use the options directly without `-o` So that most options will work fine. Split the Options by `,` then split by `=` as key and value. If value == "" then use as `--KEY` else use as `--KEY=VALUE`.

@aravindavk yes I also thought about the same but felt it'll be a half baked solution and just reverted for now.

So, want me to invest time on this and implement maximum possible options?

aravindavk commented:
> > Instead of revert, Change the code to use the options directly without `-o` So that most options will work fine. Split the Options by `,` then split by `=` as key and value. If value == "" then use as `--KEY` else use as `--KEY=VALUE`.
> 
> @aravindavk yes I also thought about the same but felt it'll be a half baked solution and just reverted for now.

It will be good fix with most options working. Only a few special cases wont work(Fuse options). That can be added later.

> 
> So, want me to invest time on this and implement maximum possible options?

Not now. This fix should work for the safe cases.

aravindavk commented on csi/volumeutils.py:
```diff
@@ -1135,6 +1135,19 @@ def mount_glusterfs_with_host(volname, mountpoint, hosts, options=None, is_clien
+    #     for opt in options.split(","):
+    #         if not opt or opt == "":
+    #             break
+    #         for k,v in opt.split("="):
```
Use `opt.split("=", 1)` to avoid spliting into more parts when value contains `=`. (For example: `xlator-option=*-posix.health-check-interval=0`)

leelavg commented:
- ack to all, will get back.

leelavg commented:
@aravindavk for now I think below would be good enough, pardon the hurry, will re-visit the moana options parsing and implement them.
``` python
# bpython
bpython version 0.21 on top of Python 3.9.7 /usr/bin/python3
>>> inp='log_file="/var/log/gluster/gluster.log",no-root-squash,xlator-options=*-posix-health-check-interval=0'
>>> [f"--{opt.replace('=', ' ', 1)}" for opt in inp.split(',')]
['--log_file "/var/log/gluster/gluster.log"', '--no-root-squash', '--xlator-options *-posix-health-check-interval=0']
>>> 
```

aravindavk commented:
> @aravindavk for now I think below would be good enough, pardon the hurry, will re-visit the moana options parsing and implement them.
> 
> ```python
> # bpython
> bpython version 0.21 on top of Python 3.9.7 /usr/bin/python3
> >>> inp='log_file="/var/log/gluster/gluster.log",no-root-squash,xlator-options=*-posix-health-check-interval=0'
> >>> [f"--{opt.replace('=', ' ', 1)}" for opt in inp.split(',')]
> ['--log_file "/var/log/gluster/gluster.log"', '--no-root-squash', '--xlator-options *-posix-health-check-interval=0']

This format is wrong as args. Either use `=` to send as single element in args else make it two args.

`["--log_file", "/var/log/gluster/gluster.log", '--no-root-squash', '--xlator-options', '*-posix-health-check-interval=0']`

OR

`['--log_file="/var/log/gluster/gluster.log"', '--no-root-squash', '--xlator-options=*-posix-health-check-interval=0']`

> >>> 
> ```

aravindavk commented:
```
inp='log_file="/var/log/gluster/gluster.log",no-root-squash,xlator-options=*-posix-health-check-interval=0'
[f"--{opt}" for opt in inp.split(',')]
```
